### PR TITLE
add a --debug mode to the devtools server

### PR DESCRIPTION
**Pull request by devoncarew**
- add a --debug mode to the devtools server

As discussed by VC. The best description of the work is in the contributing doc change in this PR. Note that a large amount of changes in `lib/src/server.dart` is just a single indent change (some code is now one additional level of braces).

### CODE REVIEW

File: packages/devtools_server/lib/src/external_handlers.dart [hunk 1]
```diff
@@ -25,36 +27,73 @@ import 'usage.dart';
...
+      workingDirectory: path.join('..', 'devtools_app'),
+    ).then((Process process) {
+      // Write all flutter run process output to the server's output.
+      process.stdout.transform(utf8.decoder).listen(stdout.write);
+      process.stderr.transform(utf8.decoder).listen(stderr.write);
+
+      // Proxy all stdin to the flutter run process's input.
+      //stdin.pipe(process.stdin);
+      stdin.lineMode = false;
+      stdin.listen((event) {
```
Comment (kenzieschmoll) [inline packages/devtools_server/lib/src/external_handlers.dart]: nit: you could instead write 
```
stdin..lineMode = false
  ..listen(() {...});
```

File: packages/devtools_server/lib/src/external_handlers.dart [hunk 2]
```diff
@@ -25,36 +27,73 @@ import 'usage.dart';
...
+
+    // ignore: unawaited_futures
+    Process.start(
+      'flutter',
+      ['run', '-d', 'web-server', '--web-port=9101'],
+      workingDirectory: path.join('..', 'devtools_app'),
+    ).then((Process process) {
+      // Write all flutter run process output to the server's output.
+      process.stdout.transform(utf8.decoder).listen(stdout.write);
+      process.stderr.transform(utf8.decoder).listen(stderr.write);
```
Comment (kenzieschmoll) [inline packages/devtools_server/lib/src/external_handlers.dart]: you could instead write
```
process..stdout[...]
  ..stderr[...];
```

File: packages/devtools_server/lib/src/external_handlers.dart [hunk 3]
```diff
@@ -25,36 +27,73 @@ import 'usage.dart';
...
+      stdin.lineMode = false;
+      stdin.listen((event) {
+        process.stdin.add(event);
+      });
+
+      // Exit when the flutter run process exits.
+      process.exitCode.then(exit);
+    });
+
+    debugProxyHandler = proxyHandler(Uri.parse('http://localhost:9101/'));
```
Comment (kenzieschmoll) [inline packages/devtools_server/lib/src/external_handlers.dart]: pull 9101 out into a constant and then inject it using string interpolation here and above. That way if we ever change the port number we make sure we change it in both places.

File: packages/devtools_server/lib/src/external_handlers.dart [hunk 4]
```diff
@@ -25,36 +27,73 @@ import 'usage.dart';
...
   // The packages folder is renamed in the pub package so this handler serves
   // out of the `pack` folder.
-  final packagesDir = path.join(packageDir, 'build', 'pack');
-  final packHandler = createStaticHandler(
-    packagesDir,
-    defaultDocument: 'index.html',
-  );
+  Handler packHandler;
+  if (!debugMode) {
+    packHandler = createStaticHandler(
```
Comment (kenzieschmoll) [inline packages/devtools_server/lib/src/external_handlers.dart]: do we need to set `packHandler` for the debug case?

File: packages/devtools_server/lib/src/server.dart [hunk 5]
```diff
@@ -120,6 +128,7 @@ Future<HttpServer> serveDevToolsWithArgs(List<String> arguments,
   final bool enableNotifications = args[argEnableNotifications];
   final port = args[argPort] != null ? int.tryParse(args[argPort]) ?? 0 : 0;
   final bool headlessMode = args[argHeadlessMode];
+  final bool debugMode = args['debug'];
```
Comment (kenzieschmoll) [inline packages/devtools_server/lib/src/server.dart]: perhaps make `debug` a constant like the other keys (e.g. argHeadlessMode)

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -55,6 +55,26 @@ To test release performance:
 You can also use `-d headless-server`, which will start a headless server that serves the HTML
 files for the DevTools Flutter app.
 
+## Development (DevTools server + DevTools web app)
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -55,6 +55,26 @@ To test release performance:
...
+
+To develop with a workflow that excercises the DevTools server <==> DevTools client connection,
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply CONTRIBUTING.md]: done

File: packages/devtools_server/lib/src/external_handlers.dart (hunk 2 above)
Comment (devoncarew) [reply packages/devtools_server/lib/src/external_handlers.dart]: changed

File: packages/devtools_server/lib/src/external_handlers.dart (hunk 1 above)
Comment (devoncarew) [reply packages/devtools_server/lib/src/external_handlers.dart]: and done

File: packages/devtools_server/lib/src/external_handlers.dart (hunk 3 above)
Comment (devoncarew) [reply packages/devtools_server/lib/src/external_handlers.dart]: sounds good - moved to using a const value

File: packages/devtools_server/lib/src/external_handlers.dart (hunk 4 above)
Comment (devoncarew) [reply packages/devtools_server/lib/src/external_handlers.dart]: No - the flutter run web compilation server will handle all that for us.

File: packages/devtools_server/lib/src/server.dart (hunk 5 above)
Comment (devoncarew) [reply packages/devtools_server/lib/src/server.dart]: will do

**Comment by devoncarew** [comment]
I have a few tests to fix up here -

**Comment by devoncarew** [comment]
```
The following NoSuchMethodError was thrown building DefaultAssetBundle:
The method 'getValue' was called on null.
Receiver: null
Tried calling: getValue("ui.darkMode")

The relevant error-causing widget was:
  DefaultAssetBundle
  file:///Users/devoncarew/projects/flutter/devtools/packages/devtools_app/test/flutter/integration/dart_cli_profile_test.dart:53:19

When the exception was thrown, this was the stack:
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      new PreferencesController (package:devtools_app/src/flutter/preferences.dart:14:13)
#2      new DevToolsAppState (package:devtools_app/src/flutter/app.dart:53:45)
#3      DevToolsApp.createState (package:devtools_app/src/flutter/app.dart:39:39)
#4      new StatefulElement (package:flutter/src/widgets/framework.dart:4591:25)
#5      StatefulWidget.createElement (package:flutter/src/widgets/framework.dart:894:38)
...     Normal element mounting (7 frames)
#12     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3446:14)
#13     Element.updateChild (package:flutter/src/widgets/framework.dart:3211:20)
#14     RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:1148:16)
#15     RenderObjectToWidgetElement.update (package:flutter/src/widgets/binding.dart:1126:5)
#16     RenderObjectToWidgetElement.performRebuild (package:flutter/src/widgets/binding.dart:1140:7)
#17     Element.rebuild (package:flutter/src/widgets/framework.dart:4218:5)
#18     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2627:33)
#19     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:883:20)
#20     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:284:5)
#21     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1108:15)
#22     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1047:9)
#23     LiveTestWidgetsFlutterBinding.handleDrawFrame (package:flutter_test/src/binding.dart:1344:13)
#24     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:963:5)
#28     _invoke (dart:ui/hooks.dart:261:10)
#29     _drawFrame (dart:ui/hooks.dart:219:3)
(elided 3 frames from dart:async)
```

**Comment by devoncarew** [comment]
> I have a few tests to fix up here -

It looks like several were just timing dependent.

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply CONTRIBUTING.md]: not sure why, but this change isn't showing up  for me. Did you commit this change? same for below

[the remark this answers is not in the document]
Comment (devoncarew) [reply CONTRIBUTING.md]: I'll double check

[the remark this answers is not in the document]
Comment (devoncarew) [reply CONTRIBUTING.md]: (I'd forgotten to commit)

**Comment by devoncarew** [comment]
This PR seems to make the tests more flaky, but I don't see why that would be.

https://travis-ci.org/github/flutter/devtools/jobs/667455490?utm_medium=notification&utm_source=github_status

**Comment by devoncarew** [comment]
Landing this, but if the bots are less stable afterwards we can revert.

**Pull request merged**